mne/gui/_vol_stc.py

```diff
@@ -0,0 +1,790 @@
+            scalars = np.array(np.where(np.isnan(self._stc_img), 0., 1.))
+            spacing = np.diag(self._src_vox_ras_t)[:3]
+            origin = self._src_vox_ras_t[:3, 3] - spacing / 2.
+            center = 0.5 * self._stc_range - self._stc_min
+            self._grid, self._grid_mesh, self._volume_pos, self._volume_neg = \
+                self._renderer._volume(
+                    dimensions=src_shape, origin=origin,
+                    spacing=spacing,
+                    scalars=scalars.flatten(order='F'),
+                    surface_alpha=self._alpha,
+                    resolution=0.4, blending='mip', center=center)
+            self._volume_pos_actor = self._renderer.plotter.add_actor(
+                self._volume_pos, render=False)[0]
+            self._volume_neg_actor = None if self._volume_neg is not None \
+                else self._renderer.plotter.add_actor(
+                    self._volume_neg, render=False)[0]
+            _, grid_prop = self._renderer.plotter.add_actor(
+                self._grid_mesh, render=False)
+            grid_prop.SetOpacity(0.1)
+            self._scalar_bar = self._renderer.scalarbar(
+                source=self._volume_pos_actor, n_labels=8, color='black',
+                bgcolor='white', label_font_size=10)
+            self._scalar_bar.SetOrientationToVertical()
+            self._scalar_bar.SetHeight(0.6)
+            self._scalar_bar.SetWidth(0.05)
+            self._scalar_bar.SetPosition(0.02, 0.2)
+
+        self._update_cmap()  # must be called for volume to render properly
+
+        if show:
+            self.show()
+
+    def _get_src_coord(self):
+        """Get the current slice transformed to source space."""
+        return tuple(np.round(_coord_to_coord(
+            self._current_slice, self._vox_scan_ras_t,
+            self._src_scan_ras_vox_t)).astype(int))
+
+    def _pick_stc_epoch(self):
+        """Select the source time course epoch based on the parameters."""
+        if self._epoch_idx == 'Average':
+            if self._data.dtype == np.int16:
+                stc_data = self._data.mean(axis=0).astype(np.int16)
+            else:
+                stc_data = self._data.mean(axis=0)
+        elif self._epoch_idx == 'Average Power':
+            if self._data.dtype == COMPLEX_DTYPE:
+                stc_data = np.sum(_int_complex_conj(
+                    self._data) // self._data.shape[0], axis=0)
+            else:
+                stc_data = (self._data * self._data.conj()).real.mean(axis=0)
+        elif self._epoch_idx == 'ITC':
+            if self._data.dtype == COMPLEX_DTYPE:
+                stc_data = self._data['re'].astype(np.complex64) + \
+                    1j * self._data['im']
+                stc_data = np.abs((stc_data / np.abs(stc_data)).mean(
+                    axis=0))
+            else:
+                stc_data = np.abs((self._data / np.abs(self._data)).mean(
+                    axis=0))
```

Comment:
would there be a way to avoid having some method logic in the GUI? basically we could have other stuff than ITC, or average_power (think real part of the coherence etc.) and I am a bit worried to see this implemented in a GUI code

just tell me what you think

Reply:
The reason to implement it here is that it's really slow using float and it's relatively easy and much faster to use integers. There's not a cleanly abstracted function for ITC in `tfr.py` anyway so that would need to be refactored. We could make `_itc` and `_itc_int` functions in `tfr.py` if you want, I just was trying to keep the number of files touched low.

Reply:
It doesn't look like the most straightforward abstraction/refactor from here since it's looped

```
# Inter-trial phase locking is apparently computed per taper...
if 'itc' in output:
    plf = np.zeros((n_freqs, n_times), dtype=np.complex128)

# Loop across epochs
for epoch_idx, tfr in enumerate(coefs):
    # Transform complex values
    if output in ['power', 'avg_power']:
        tfr = (tfr * tfr.conj()).real  # power
    elif output == 'phase':
        tfr = np.angle(tfr)
    elif output == 'avg_power_itc':
        tfr_abs = np.abs(tfr)
        plf += tfr / tfr_abs  # phase
        tfr = tfr_abs ** 2  # power
    elif output == 'itc':
        plf += tfr / np.abs(tfr)  # phase
        continue  # not need to stack anything else than plf

    # Stack or add
    if ('avg_' in output) or ('itc' in output):
        tfrs += tfr
    elif output in ['complex', 'phase'] and method == 'multitaper':
        tfrs[taper_idx, epoch_idx] += tfr
    else:
        tfrs[epoch_idx] += tfr
```